Add global option to skip households on simulation failure

i-am-sijia opened the pull request:
This PR addresses issue #892.

**Status:** Draft - core functionality implemented, remaining: logging and documentation. Update: logging and documentation are added.


## Summary
Adds a new `skip_failed_choices` setting that allows ActivitySim to skip households that cause choice model failures instead of being masked and forced a choice by the `overflow_pretection` or crashing the entire simulation. This improves robustness for large-scale production runs where occasional edge cases shouldn't halt the model.

## Key Changes

**Core Logic (`activitysim/core/`)**
- Modified `logit.py` to temporarily mark failed choices with `-99` and track skipped households in state
- Added functionality in `state.py` to dynamically drop data that belong to the skipped household. Further updated the logic to only trigger this when new households are being skipped.
- Updated `simulate.py` and `interaction_sample_simulate.py` to handle failed choices
- Enhanced `report_bad_choices()` to log skipped households
- Added code in `simulate.py` to automatically print out debug tracing for one failed choice per model component, limited to simple simulate models only.

**Configuration**
- Added `skip_failed_choices: bool = True` setting in `top.py`
- Added `fraction_of_failed_choices_allowed = 0.1` setting in `top.py`. Users can optionally specify different values in the settings.yaml to set the threshold of failed/skipped households (relative to sample households) above which the run should terminate.
- Added code in `cli/run.py` to print out a summary of skipped households by model component at the end of the run.

**Model Updates (`activitysim/abm/models/`)**
- Retained `household_id` in chooser columns across location choice models for tracking
- Modified shadow price calculations to exclude skipped households
- Updated trip matrices to adjust trip matrix expansion factor. The adjustment is based on the remaining households' weight vs the initial sampled households' weight.
- Added null trip mode handling in trip mode choice and school escorting

**Testing**
- Added new unit test `test/skip_failed_choices` to deliberately trigger failures and validates households being skipped.
- Added tests with different threshold of skipped households.

**Documentation**
- Added documentation in `docs/users-guide/modelsetup.rst` and `docs/users-guide/ways_to_run.rst`.

## Usage
```yaml
# settings.yaml
skip_failed_choices: true #default
fraction_of_failed_choices_allowed: 0.1 #default
```

Access skipped household info via:
- `state.get("num_skipped_households", 0)`
- `state.get("skipped_household_ids", dict())`

Final count logged at end of simulation in the `activitysim.log`.

## Notes
- Defaults to `True` for production robustness
- Disables overflow protection when active to surface underlying issues
- Temporary code adds `household_id` to chooser columns (remove when PR #1017 merges)

i-am-sijia committed:
skip hh in simple simulate

i-am-sijia committed:
retain `household_id` by default in the drop_unused_columns()

we need it for debug tracing and skipping failed hh

i-am-sijia committed:
skip choice for interaction sample simulate

i-am-sijia committed:
don't log hh id when bad choice df is empy

i-am-sijia committed:
add hh id to user specified `chooser_columns` when `chooser_columns` exist

i-am-sijia committed:
blacken

i-am-sijia committed:
update test utils to probs raises

i-am-sijia committed:
[skip ci] shadow price calc without failed choices

i-am-sijia committed:
[skip ci] special case in trip mode choice

i-am-sijia committed:
[skip ci] adjust sample_rate

i-am-sijia committed:
[skip ci] add final logging of num hh skipped

i-am-sijia committed:
unit test

i-am-sijia committed:
update CI

i-am-sijia commented:
For components where we intentionally allow some choices to fail without being skipped, the `allow_zero_probs` option is already set to `True`, which exempts them from this new feature. Therefore, it may not be necessary to introduce this as a component-level setting. In addition, keeping it as a global setting helps reduce manual errors, avoid confusion, and prevent inconsistencies across the implementation.

jpn-- commented on activitysim/abm/models/util/school_escort_tours_trips.py:
```diff
@@ -1043,6 +1043,10 @@ def force_escortee_trip_modes_to_match_chauffeur(state: workflow.State, trips):
 
+    # trip_mode can be na if the run allows skipping failed choices and the trip mode choice has failed
+    if state.settings.skip_failed_choices:
+        return trips
```
Rather than just being OK here, code should:
1. emit a warning that N choices have failed, and
2. check if N exceeds configured failure thresholds (e.g. not more than 1% of all choices), in which case an error should be raised.

i-am-sijia replied:
Added the warning for item 1. For item 2, I implemented the check in core.workflow.state.update_table(), immediately after the households are being dropped to make sure it raises an error when N (or the weighted N) exceeds the threshold. Implementing the check in the state instead of in model components makes the code base clean, i.e., less code duplication. The global parameter of the threshold is added in core.configuration.top.fraction_of_failed_choices_allowed.

jpn-- commented on activitysim/abm/models/location_choice.py:
```diff
@@ -1072,6 +1087,23 @@ def iterate_location_choice(
+            how="left",
+        )
+        if len(choices_df) > 0:
+            choices_df = choices_df[
```
Should check here how many choices are getting dropped, and
1. emit a warning that N choices have failed, and
2. check if N exceeds configured failure thresholds (e.g. not more than 1% of all households), in which case an error should be raised.

i-am-sijia replied:
Both the warning and the check have been implemented in `core.workflow.state.update_table()`. I implemented them in the state.py instead of in model components to make the code base clean, i.e., less code duplication.

jpn-- commented on activitysim/abm/models/trip_matrices.py:
```diff
@@ -307,6 +307,12 @@ def write_trip_matrices(
 
+    # print out number of households skipped due to failed choices
+    if state.settings.skip_failed_choices:
+        logger.info(
```
logging level for this should be higher than "info"; at least warning if not "error" level.

i-am-sijia replied:
I've moved the summary of number of skipped households, total and total by model component, to `cli.run.run()`, which is a better home than trip_matrices.py.

jpn-- commented on activitysim/abm/models/trip_matrices.py:
```diff
@@ -393,6 +399,21 @@ def write_matrices(
+                f"Adjusting household expansion weights in {hh_weight_col} to account for {state.get('num_skipped_households', 0)} skipped households."
+            )
+            # adjust the hh expansion weights to account for skipped households
+            adjustment_factor = state.get_dataframe("households").shape[0] / (
```
Should the adjustment to expansion weights be based not on the number of skipped households, but instead on the original expansion weight total and the total expansion weight of the households we dropped?  E.g. if we drop a household with a giant expansion weight that's a bigger thing than dropping a household with a smaller one.

i-am-sijia replied:
I've updated the adjustment to expansion weights to be based on the weight of households instead of counts. I agree with you in theory that it should be based on weights. However, I don't think it will make a difference in application. Here's why. In simulation, 99% of the time ActivitySim has the same [sample rate](https://github.com/ActivitySim/activitysim/blob/aa58cc453e1a0ff5c7d5498776f2f594b6085658/activitysim/abm/tables/households.py#L98-L104) for all households and the same expansion weight (calculated as [1/sample rate](https://github.com/ActivitySim/activitysim/blob/aa58cc453e1a0ff5c7d5498776f2f594b6085658/activitysim/abm/models/trip_matrices.py#L415-L418)), this is especially true when ActivitySim uses an input synthetic population with weight of 1 for each record implicitly built in. In contrast to simulation, in estimation, the input households are very likely to have different weights, however, we don't use weights when estimating a model, and `trip_matrices` step is usually skipped in the estimation mode.

jpn-- commented on activitysim/abm/models/trip_mode_choice.py:
```diff
@@ -382,6 +389,11 @@ def trip_mode_choice(
     locals_dict = {}
     locals_dict.update(constants)
+    if state.settings.skip_failed_choices:
+        mask_skipped = trips_merged["household_id"].isin(
```
mask_skipped should be the same for trips_df and trips_merged, right?  We only need to calculate the mask once.

i-am-sijia replied:
Yes, you're right. Updated.

jpn-- commented on activitysim/core/configuration/top.py:
```diff
@@ -781,6 +781,13 @@ def _check_store_skims_in_shm(self):
+
+    .. versionadded:: 1.6
+    """
+
```
Need additional setting[s] to set thresholds for how many skips are OK and when it's too many and should be an error.

i-am-sijia replied:
Added `fraction_of_failed_choices_allowed` as a global parameter and implemented the checking in `core.workflow.state.update_table()`

jpn-- commented on activitysim/core/logit.py:
```diff
@@ -87,6 +88,27 @@ def report_bad_choices(
+        state.set("num_skipped_households", num_skipped_households)
+        state.set("skipped_household_ids", skipped_household_ids)
+
+        logger.debug(
```
logger level here should be `warn` not `debug`

i-am-sijia replied:
Updated to `warning` for all loggings of skipping households.

jpn-- reviewed (changes requested):
Mostly looks fine.  We discussed (possibly while @i-am-sijia was out) that the skipping should not be unlimited; the idea is we want to skip problems when problems are few and probably not seriously consequential.  If the problems are many there's little benefit of allowing a simulation to run to completion.

i-am-sijia committed:
track and summarize failed households by component

i-am-sijia committed:
`skipped_household_ids` is now a dictionary

i-am-sijia committed:
print more loggings about trips missing modes

i-am-sijia committed:
adjust trip expansion factor using hh weight instead of count

i-am-sijia committed:
update state tables only when new households are skipped

i-am-sijia committed:
put a limit on num of households allowed to skip

i-am-sijia commented on activitysim/abm/models/util/school_escort_tours_trips.py:
```diff
@@ -1043,6 +1043,10 @@ def force_escortee_trip_modes_to_match_chauffeur(state: workflow.State, trips):
     )
 
+    # trip_mode can be na if the run allows skipping failed choices and the trip mode choice has failed
+    if state.settings.skip_failed_choices:
```
Added warning and additional assertion.

i-am-sijia commented on activitysim/core/simulate.py:
```diff
@@ -1485,11 +1487,15 @@ def eval_nl(
             state,
             no_choices,
             base_probabilities,
+            skip_failed_choices,
```
addressed

i-am-sijia commented on activitysim/core/workflow/state.py:
```diff
@@ -939,6 +939,62 @@ def add_table(
+        # TODO: do not update tables if no new households were skipped
+        # right now it is updating every time which is inefficient
+        if self.get("num_skipped_households", 0) > 0:
+            self.update_table()
```
Thanks copilot. As I noted in the TODO, I've added the code to update tables only when new households are being skipped.

i-am-sijia commented on activitysim/core/util.py:
```diff
@@ -683,6 +683,9 @@ def drop_unused_columns(
 
     unique_variables_in_spec |= set(additional_columns or [])
 
+    # always keep household_id
+    unique_variables_in_spec.add("household_id")
```
This is fine, no changes needed. As in line 737 below, we are keeping columns in chooser table IF they exist in the `unique_variables_in_spec`.

i-am-sijia commented on activitysim/core/workflow/state.py:
```diff
@@ -939,6 +939,62 @@ def add_table(
+        if self.get("num_skipped_households", 0) > 0:
+            self.update_table()
+
+    def update_table(self, name: str = None):
```
This method has been substantially updated to implement a few additional features.

i-am-sijia committed:
auto debug tracing newly failed choice

i-am-sijia committed:
update test

i-am-sijia committed:
Add documentation to user guide

i-am-sijia committed:
Merge branch 'main' into skip-failed-hh

i-am-sijia commented:
@jpn-- Thank you for your review and comments! I've made further updates. Primary changes include:

- Added a `fraction_of_failed_choices_allowed` parameter that users can optionally specify in the `settings.yaml` to set the **threshold** of failed/skipped households (relative to sample households) above which the run should terminate. I updated the test to incorporate this as well.
- Added code to summarize and print out number of skipped households **by model component** at the end of the run.
- Added code to update the state tables only when there're **new** households skipped, improving efficiency.
- Updated code to adjust trip matrix expansion factor using household **weight** instead of count. For this item, please see my response comment to your original comment.
- As requested by @bhargavasana, I added a feature to automatically trace one failed choice per model component. This turned out to be more complex than I expected. I couldn't find an clean way to do it without significant refactoring of the current tracing mechanism and the utility calculation. I managed to implement a workaround for simple simulate models by dynamically adding trace IDs to the tracing object, and re-simulate a single failed choice to trace the utility calculation with minimal runtime impact runtime. For the interaction simulate models, users can manually add the failed household ID to `trace_hh_id` for debug tracing.
- Added documentation in the user guide.

dhensle commented on docs/users-guide/ways_to_run.rst:
```diff
@@ -283,3 +283,35 @@ With the set of output CSV files, the user can trace ActivitySim calculations in
+Skip Failed Choices
+______________________
+
+By default, ActivitySim will skip any failed choices during model execution, i.e., ``skip_failed_choices`` is set to ``True``. 
```
I would argue that we want the default behavior set to False.  Having non-expert users just running ActivitySim and dropping households could really confuse folks and cause outputs to not make sense.  Imagine, for example, a planning practitioner is trying to run a parking cost scenario -- they update the parking costs but miss a zone, that zone propagates NAs, and all of a sudden, they have 10% less households in the output file than the input file.  Run says it succeeded and they think all is groovy.  But their revenue forecast is now off by 10% and they don't even know!

bhargavasana replied:
I support this - it might be better have default set to False.

i-am-sijia replied:
As discussed on the engineering call today, this is ultimately a product decision that the product & community team will make. I'm comfortable with either direction. I did want to add a few considerations from similar / different angle:

1. Enabling `skip_failed_choices` could cause existing model run results to change, because it would surface choice failures that are currently masked by the `overflow_protection` feature. I recognize the value of preserving agencies' existing implementations.
2. When `skip_failed_choices` is set to False, the 10% households in David's example still make a choice because `overflow_protection` kicks in and masks the failure. Those households will very likely not be responsive to parking cost changes. The practitioner wouldn't know about this behavior either, and may be confused about why the model appears insensitive. This is a tradeoff worth considering.
3. David is right that the output household is fewer than the input, but that's the intended design, just like how ActivitySim runs on a sample of households. This is reflected in the adjusted `sample_rate` column in the output households, and the resulting trip tables and matrices are scaled up accordingly.

bhargavasana replied:
#1052 created to get input from @ActivitySim/product.

dhensle commented on activitysim/abm/models/trip_matrices.py:
```diff
@@ -90,6 +90,34 @@ def write_trip_matrices(
+    # based on household_sample_size / total_household_count, and therefore is the same for all households.
+    # In estimation, the `sample_rate` may vary by household, but weights are not used in estimation, and write_trip_matrices is not called during estimation.
+    # But we still try to cover both cases (when rates are the same vs when they vary) here for consistency.
+    hh_weight_col = model_settings.HH_EXPANSION_WEIGHT_COL
```
Note that the sample_rate is set to only 3 digits of rounding.  This means that for cases where we are running, say, only 100 households to test some spec development on a region with 1.2M households, you will likely get a divide by 0 error here. I suggest we increase the rounding to something much more than 3 and also put in a bit of divide-by-zero protection in this block.

i-am-sijia replied:
That's a good point. I will add that. Thanks!

jpn-- replied:
@i-am-sijia I think the divide-by-zero problem has been reduced by setting the rounding to 7 instead of 3, but the problem still potentially exists here; the sample_rate could still in some cases be zero.  Can we do a np.where or similar to solve it?

Also: why are we rounding this at all?  It seems to me it doesn't hurt us to just store and use the value as a float.

i-am-sijia replied:
I think skip rounding and directly use the value is the better solution. `sample_rate` is a `float64` regardless of the number of decimals.

dhensle commented on activitysim/core/workflow/state.py:
```diff
@@ -938,8 +938,181 @@ def add_table(
+
+                # Check if we've exceeded the allowed fraction of skipped households
+                # Use weighted households if expansion weight column exists, otherwise use counts
+                if "sample_rate" in df.columns:
```
See above comment about sample rate.  I think it applies here too.

dhensle commented on test/skip_failed_choices/test_skip_failed_choices.py:
```diff
@@ -0,0 +1,160 @@
+    assert "exceeds the allowed threshold of" in str(excinfo.value)
+
+
+def test_skip_failed_choices_high_threshold_prototype_mtc(state):
```
Can we add an additional check here to make sure a trace gets created for the last household that was skipped for the auto ownership and trip mode choice models?

i-am-sijia replied:
added.

dhensle commented on docs/users-guide/ways_to_run.rst:
```diff
@@ -283,3 +283,35 @@ With the set of output CSV files, the user can trace ActivitySim calculations in
+When ``skip_failed_choices`` is enabled, ActivitySim will automatically perform debug tracing for one of the failed households within each 
+model step where failed choices occur. The trace files will be saved in the output/trace directory with folders suffixed by 
+``_resimulate``. This automatic tracing feature allows users to easily investigate the reasons behind the failed choices without needing to 
+manually specify trace IDs. This feature is partially implemented for simple simulate models, and is not yet available for interaction_simulate models.
```
What do you mean by "partially implemented for simple simulate models" here?

i-am-sijia replied:
It is fully implemented for simple simulate. Corrected wording.

dhensle reviewed (commented):
Looks good! Just a couple of relatively minor suggestions.

i-am-sijia committed:
update doc

i-am-sijia committed:
increase sample rate decimals

i-am-sijia committed:
handle zero `sample_rate` when calculating weights

i-am-sijia committed:
black

i-am-sijia committed:
[skip ci] confirm trace files are generated for skipped hh

i-am-sijia committed:
Revert "increase sample rate decimals"

This reverts commit a734c16eb5beb1ddbe88bb575c02111c34442f50.

i-am-sijia committed:
add note for not rounding sample rate

i-am-sijia committed:
write trip matrices even when `sample_rate` is all zero

To pass this single hh test: https://github.com/ActivitySim/activitysim/actions/runs/23202170408/job/67427319813?pr=1023

i-am-sijia commented:
@jpn-- @dhensle , I've implemented divide-by-zero protection for `sample_rate` as discussed at https://github.com/ActivitySim/meeting-notes/issues/79. Please give it a review and let me know if you have any comments.

Software Behavior:
1) **When `sample_rate==0`  for _all_ households**
This can occur when simulating a very small household sample (e.g., ~20 out of 2M) and a computed `sample_rate` is rounded down to zero. 
In this case, the adjustment factor is calculated using household counts instead of weights.

2) **When `sample_rate==0` for _some_ households (partial zeros)**
This can occur when `sample_rate` is provided in the input household data and includes zeros (unclear whether intended or a data error).
In this case, the adjustment factor is computed using only non-zero-sample_rate households (weights from non-zero-sample_rate households only; zero-sample_rate households are skipped). Trips from zero-sample_rate households are skipped when writing out trip matrices.

As discussed, we will remove the rounding of `sample_rate` in a separate PR, with updates to our testing data. I've reverted rounding to 3 decimals as it originally was, to pass test validations.

dhensle reviewed (approved):
Looks good, thanks Sijia!
